### PR TITLE
Use selfbalance for ``address(this).balance``.

### PR DESCRIPTION
**Pull request by chriseth**
References https://github.com/ethereum/solidity/issues/7310

### CODE REVIEW

File: libsolidity/codegen/ExpressionCompiler.cpp [hunk 1]
```diff
@@ -1229,6 +1229,29 @@ bool ExpressionCompiler::visit(MemberAccess const& _memberAccess)
 						utils().leftShiftNumberOnStack(224);
 						return false;
 					}
+	// Another special case for `address(this).balance`. Post-Istanbul, we can use the selfbalance
+	// opcode.
+	// TODO also add optimizer rule `balance(address) -> selfbalance`.
```
Comment (chriseth) [inline libsolidity/codegen/ExpressionCompiler.cpp]: This needs information about the dialect together with the rules (or applicability function?).

File: test/libsolidity/SolidityExpressionCompiler.cpp [hunk 2]
```diff
@@ -617,6 +617,25 @@ BOOST_AUTO_TEST_CASE(gas_left)
...
+	)";
+
+	bytes code = compileFirstExpression(sourceCode, {}, {});
+
+	if (dev::test::Options::get().evmVersion() == EVMVersion::istanbul())
+	{
+		bytes expectation({uint8_t(Instruction::SELFBALANCE)});
+		BOOST_CHECK_EQUAL_COLLECTIONS(code.begin(), code.end(), expectation.begin(), expectation.end());
+	}
+}
```
Comment (erak) [inline test/libsolidity/SolidityExpressionCompiler.cpp:637]: Does it make sense to check that `BALANCE` together with `ADDRESS` is used for pre-Istanbul here? Couldn't find a test which does that check.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (erak) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Shall we create an issue which tracks this?

**Comment by erak** [comment]
Tests are failing due to a weird Circle glitch(?). We can try to retrigger the tests.

File: libsolidity/codegen/ExpressionCompiler.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/codegen/ExpressionCompiler.cpp]: Created https://github.com/ethereum/solidity/issues/7542

**Review by erak — APPROVED** [review]

**Pull request merged**